crates/app-host/src/block_tracer.rs

```diff
@@ -0,0 +1,78 @@
+
+use anyhow::Result;
+
+use scroll_executor::BlockTrace;
+use tokio::runtime::Runtime;
+use std::sync::Arc;
+use l2_client::L2Client;
+
+pub struct BlockTracer {
+    l2_client: Arc<l2_client::L2Client>,
+    rt: Runtime,
+}
+
+impl BlockTracer {
+    pub fn new() -> Self {
+        todo!()
+    }
+
+    pub async fn get_block_traces(&self, blocks: Vec<u64>) -> Result<Vec<BlockTrace>> {
+        let mut handles = Vec::with_capacity(blocks.len());
+
+        for block in blocks {
+            let l2_client = self.l2_client.clone();
+            // todo: use a singleton tokio runtime with limited worker to spawn task
+            let handle = self.rt.spawn(async move {
+                BlockTracer::get_block_trace(l2_client, block).await
+            });
+            handles.push(handle);
+        }
+
+        let mut block_traces = Vec::with_capacity(handles.len());
+        for handle in handles {
+            let block_trace = handle.await?;
+            block_traces.push(block_trace)
+        }
+        Result::Ok(block_traces)
+    }
+
+    async fn get_block_trace(l2_client: Arc<L2Client>, block: u64) -> BlockTrace {
```

Comment:
```suggestion
    async fn get_block_trace(l2_client: Arc<L2Client>, block: u64) -> Result<BlockTrace> {
```

If it can break, it's better to return `Result`

Reply:
this is by intention since I plan to handle error and retry in this method. The block_trace is a must-have thing which causes the whole process being blocked.  I plan to handle the error here instead throw it out